gadgets/src/range.rs

```diff
@@ -0,0 +1,153 @@
+//! Specialized version to replace LtChip for u32/u64/u128 check
+//!
+//! only constrain input expressions to be in given range
+//! do not output the result
+//!
+//! u8 check can either use U8Table or group every 2 values and lookup U16Table
+//! u16 check can use U16Table directly
+
+use crate::util::Expr;
+use eth_types::Field;
+use halo2_proofs::{
+    arithmetic::FieldExt,
+    circuit::{Chip, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, TableColumn, VirtualCells},
+    poly::Rotation,
+};
+
+/// Instruction that the Range chip needs to implement.
+pub trait UIntRangeCheckInstruction<F: FieldExt, const N_2BYTE: usize, const N_EXPR: usize> {
+    /// Assign the expr and u16 le repr witnesses to the Comparator chip's region.
+    fn assign(
+        &self,
+        region: &mut Region<F>,
+        offset: usize,
+        values: [F; N_EXPR],
+    ) -> Result<(), Error>;
+}
+
+/// Config for the UIntRange chip.
+///
+/// `N_2BYTE` is size of range in (u16) 2-byte.
+/// for u32, N_2BYTE = 2; for u64, N_2BYTE = 4; for u128, N_2BYTE = 8
+///
+/// `N_EXPR` is the number of lookup expressions to check.
+#[derive(Clone, Copy, Debug)]
+pub struct UIntRangeCheckConfig<F, const N_2BYTE: usize, const N_EXPR: usize> {
+    /// Denotes the little-endian representation of expression in u16.
+    pub u16_repr: [Column<Advice>; N_2BYTE],
+    /// Denotes the u16 lookup table.
+    pub u16_table: TableColumn,
+    _marker: std::marker::PhantomData<F>,
+}
+
+/// Chip that checks if expressions are in range.
+#[derive(Clone, Debug)]
+pub struct UIntRangeCheckChip<F, const N_2BYTE: usize, const N_EXPR: usize> {
+    config: UIntRangeCheckConfig<F, N_2BYTE, N_EXPR>,
+}
+
+impl UIntRangeCheckChip<(), 0, 0> {
+    /// constant value of `N_2BYTE` for u32 check
+    pub const SIZE_U32: usize = 2;
+    /// constant value of `N_2BYTE` for u64 check
+    pub const SIZE_U64: usize = 4;
+    /// constant value of `N_2BYTE` for u128 check
+    pub const SIZE_U128: usize = 8;
+}
+
+impl<F: Field, const N_2BYTE: usize, const N_EXPR: usize> UIntRangeCheckChip<F, N_2BYTE, N_EXPR> {
+    /// Configures the range chip.
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        q_enable: impl FnOnce(&mut VirtualCells<F>) -> Expression<F> + Clone,
+        expressions: impl FnOnce(&mut VirtualCells<F>) -> [Expression<F>; N_EXPR],
```

Comment:
(optimization)

This chip takes columns from the circuit and add equality constraints.

Instead, the chip could _return_ the expressions `acc`. Then the cells in the circuit can be removed (everything passed in this `expressions` argument).